demo/admin/src/common/apollo/apolloClient.ts

```diff
@@ -0,0 +1,19 @@
+import { ApolloClient, ApolloLink } from "@apollo/client";
+import { includeInvisibleContentContext } from "@comet/admin-cms";
+import { createAuthorizationLink } from "@comet/react-app-auth";
+
+import { authorizationConfig } from "../authorization/authorizationConfig";
+import { refreshHandler } from "../authorization/refreshHandler";
+import { inMemoryCache } from "./inMemoryCache";
+import { httpLink } from "./links/httpLink";
+
+const link = ApolloLink.from([
+    includeInvisibleContentContext,
+    createAuthorizationLink({ authorizationConfig, refreshHandler: refreshHandler }),
+    httpLink,
+]);
+
+export const apolloClient = new ApolloClient({
```

Comment:
(we had that discussion in another merge request, but I don't remember with one)

With this pattern you export in `common/apollo/apolloClient.ts` `apolloClient` (an instance of `ApolloClient`). Any application code can access this instance by `import { apolloClient } from "@src/common/apollo/apolloClient"`. This pattern can be convenient but has downsides:

- this won't work in library code
- it will get a mixture of accessing the ApolloProvider and the new export in the application
- this is not Inversion of Control / Dependency Injection pattern anymore

my proposal:
- reduce files (integrate `authorizationConfig` into `authorizationManager.ts`, I don't see a reason why the config needs to be exported, create `inMemoeryCache` and `httpLient` in `apolloClient.ts`)
- instead of exporting the `apolloClient` instance, export a factory `createApolloClient`
- find a solultion for refreshHandler (Reactive Var as you did for the ErrorDialog?)

Reply:
> With this pattern you export in common/apollo/apolloClient.ts apolloClient (an instance of ApolloClient). Any application code can access this instance by import { apolloClient } from "@src/common/apollo/apolloClient". This pattern can be convenient but has downsides:

how do you want to protect this kind of things? I am not sure if we must protect this.

> this won't work in library code
- as far as i know, we have no apolloClient in the  libraries, this is implemented in the application and is supposed to be only used in applications.


> it will get a mixture of accessing the ApolloProvider and the new export in the application
- if you are in react jsx, one should not use/import the apolloClient directly. Please use `useQuery`, `useApolloClient` or any other hook for that.
- if one is probably in plain js (code behind), it may make sense to import the apolloClient and query things. i think apollo is smart enough to rerender all necessary parts of the ui (when useQuery, or useApolloClient) is used in jsx.

> this is not Inversion of Control / Dependency Injection pattern anymore
- same questions as last time, what do you mean with dependency injection, we do not have dependency injection here, or do i miss something?

> my proposal:
> - reduce files (integrate authorizationConfig into authorizationManager.ts, I don't see a reason why the config needs to be exported, create inMemoeryCache and httpLient in apolloClient.ts)

why should we do that? I like it very much when everything is in a separat file. that can solve some problems, circular import problems. If you are looking for something (e.g. the inMemoryCache or an httpLink than you can look in ide or github for this file and find it very quickly.

> - instead of exporting the apolloClient instance, export a factory createApolloClient

good idea, we can do this 👍 but imho it is still in the application and does not change a lot.

> - find a solultion for refreshHandler (Reactive Var as you did for the ErrorDialog?)

can you explain what you want to achieve, but i don't get the point 🤔

Reply:
added factory function `createApolloClient`

Reply:
> same questions as last time, what do you mean with dependency injection, we do not have dependency injection here, or do i miss something?

React Context (`ApolloProvider`) is our dependency injection. Componets use useApolloClient to get the injected instance.

Reply:
> React Context (ApolloProvider) is our dependency injection. Components use useApolloClient to get the injected instance.

but there is nothing new with this? What is your suggestion?

Reply:
> how do you want to protect this kind of things?

By not exporting instances.


>> this won't work in library code

> as far as i know, we have no apolloClient in the libraries, this is implemented in the application and is supposed to be only used in applications.

true - because it won't technically work.

> if one is probably in plain js (code behind), it may make sense to import the apolloClient and query things

In such a case where you can't use React Context, you have to find a different Dependency Injection technique. Think about library code, there you'll need this technique as well.

> > reduce files (integrate authorizationConfig into authorizationManager.ts, I don't see a reason why the config needs to be exported, create inMemoeryCache and httpLient in apolloClient.ts)

> why should we do that? I like it very much when everything is in a separat file. that can solve some problems, circular import problems. If you are looking for something (e.g. the inMemoryCache or an httpLink than you can look in ide or github for this file and find it very quickly.

Splitting up into multiple files is a good idea when you have large files, but only then. I would search for the http endpoint in apolloClient.ts, not in httpLink.ts. And the configuration of authorizationManager I would look for in authorizationManager.ts



> What is your suggestion?

My suggestion is to define exporting instances as anti-pattern and find alternatives. 

`apolloClient` you changed already, but also `httpLink`, `inMemoryCache`, `authorizationManager` and `refreshHandler`. `authorizationConfig` is a bit different as it's json only.

Reply:
@nsams I've moved the authorization initialization to the App.tsx. Please review 🙂